windows/devices/uwb/UwbDeviceConnector.cxx

```diff
@@ -335,3 +444,15 @@ UwbDeviceConnector::NotificationListenerStop()
 {
     m_notificationThread.request_stop();
 }
+
+void
+UwbDeviceConnector::RegisterDeviceEventCallbacks(::uwb::UwbDeviceEventCallbacks callbacks)
+{
+    m_deviceEventCallbacks = callbacks;
+}
+
+void
+UwbDeviceConnector::RegisterSessionEventCallbacks(uint32_t sessionId, std::weak_ptr<::uwb::UwbSessionEventCallbacks> callbacks)

```

Comment:
The reason for weak_ptr on session callback is that they capture the ```this``` pointer in the lambda closure, meaning that their lifetime is completely tied to the lifetime of the UwbSession object. Thus I had the UwbSession object own the lambda, which means that UwbDeviceConnector is merely borrowing them via weak_ptr.

Regarding the device callbacks, I'm not entirely certain what the best way to approach it is, but probably using weak pointers for that would also make sense.

Reply:
To give a little color here: `UwbSession` was originally designed to allow the caller to completely control the lifetime of the callbacks. Whether or not the `UwbSession` has a set of registered callbacks has no influence on its operation, so it doesn't need to keep those callbacks alive with a strong reference. `weak_ptr` solves this perfectly since it does not itself repreent a strong reference until it is promoted (via a successful `lock()` call) to a `shared_ptr`. This way, `UwbSession` takes and holds a strong reference only for as long as it takes to invoke the callback(s) in question.

Another thing to note is that, capturing `this` in a lambda capture doesn't affect the object's (described by `this`) lifetime or ownership in any way. It simply copies the `this` pointer into the capture. Think of a lambda being implemented like this:

```
Object theObject = new Obejct(...);
void Object::LambdaExample()
{
    int argByRef = 10;
    int argByValue = 20;    
    auto func = [this, &argByRef, argByValue](){};
    m_func = std::move(func); // store the lambda for later use
}

// the compiler generates something like this for `func` above: 
struct FuncLambda {
    Object *pThis = theObject;
    int& argByRef = argByRef;
    int argByValue = argByValue;
    // fancy stuff storing the actual lambda function code and args (none here for simplicity)
};
```
This is really bad code because the lambda captures a function-local variable, `argByRef`, by reference. Once `LambdaExample()` finishes running, the memory for `funcByRef` no longer exists, yet the lambda references it. This is obviously really bad since the lambda will silently access invalid memory if invoked after the function completes. The same applies to the `this` pointer here: if that lambda was passed to another object and invoked after `theObject` is deleted, the same problem arises.

So, if you're capturing the `this` pointer in a lambda that another object is storing, you absolutely need to ensure that the captured object is not destroyed before that lambda is used. The two object's lifetimes are thus co-dependent. Unless the lifetime of the thing capturing the `this` pointer is directly tied to the lifetime of the thing storing the lambda (ie. one object wholly owns the other), then the safest way to enforce this is with strong reference counting (ie. `shared_ptr`).

All that said, I think using a `weak_ptr` here for the connector session callbacks could be a valid solution, though it does highlight the additional challenges involved using lambdas vs class instances. With a class instance represented though a (`weak/shared/unique`) pointer, it's impossible for the callbacks to reference invalid memory since the callbacks themselves are members of the class which we know is valid (via the ref count of the shared/weak/unique) pointer. With the lambda, it could have captured anything from anywhere, and even a `weak_ptr` to the class containing the lambdas doesn't tell us anything about the validity of the callbacks themselves. That doesn't mean it won't work, only that we get a less strong guarantee about their behavior. In a vacuum, I tend to prefer the class instance method since it makes the code more robust and less prone to programmer error.

Using a value-copy for the device callbacks is less clear to me, since this implies that the source of the device callbacks will always exist as long as the connector exists, and I don't think this is true, or even if it is true, it should not have to be. Any code should be able to create a connector and reigster device callbacks, not just code in `UwbDevice`.

Reply:
I think I'm going to use a weak_ptr for the device callbacks